### PR TITLE
Skip applyFilter when no predicate to push down

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java [hunk 1]
```diff
@@ -149,6 +149,10 @@ private boolean arePlansSame(FilterNode filter, TableScanNode tableScan, PlanNod
         // don't include non-deterministic predicates
         Expression deterministicPredicate = filterDeterministicConjuncts(predicate);
 
+        if (TRUE_LITERAL.equals(deterministicPredicate)) {
+            return Optional.empty();
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java]: You should also return when `newDomain` is `all` and `!pruneWithPredicateExpression`. I.e. when `!pruneWithPredicateExpression` and there is some deterministic predicate that cannot be converted to a tuple domain (e.g an OR, or a function call).

TL;DR https://github.com/prestosql/presto/pull/1312

**Comment by findepi** [comment]
The commit message may need an improvement. It's already "resilient" in a sense that it doesn't fail when the predicate is empty (does it?)

**Comment by sopel39** [comment]
ac

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java [hunk 2]
```diff
@@ -177,6 +178,10 @@ private boolean arePlansSame(FilterNode filter, TableScanNode tableScan, PlanNod
                             domainTranslator.toPredicate(newDomain.simplify().transform(assignments::get))));
             constraint = new Constraint(newDomain, evaluator::isCandidate);
         }
+        else if (newDomain.isAll() && !usesLegacyTableLayouts) {
+            // for legacy layouts the only way of passing desired columns during planning is via ConnectorMetadata.getTableLayouts
+            return Optional.empty();
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java]: The comment sounds as if you'd like to return here for legacy layouts, but you're returning for non-legacy layouts. I don't understand.

What's the point of this additional condition in the `if`?
If domain is all and there is no evaluator involved, what's the point in continuing?

File: presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraTableHandle.java [hunk 3]
```diff
@@ -30,26 +30,26 @@
...
             String schemaName,
             String tableName,
-            List<CassandraPartition> partitions,
+            Optional<List<CassandraPartition>> partitions,
             String clusteringKeyPredicates)
     {
         this.schemaName = requireNonNull(schemaName, "schemaName is null");
         this.tableName = requireNonNull(tableName, "tableName is null");
-        this.partitions = ImmutableList.copyOf(requireNonNull(partitions, "partitions is null"));
+        this.partitions = requireNonNull(partitions, "partitions is null");
```
Comment (findepi) [inline presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraTableHandle.java]: `requireNonNull(partitions, "partitions is null").map(ImmutableList::copyOf)`

File: presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraSplitManager.java [hunk 4]
```diff
@@ -45,23 +46,37 @@
...
     public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, ConnectorSession session, ConnectorTableHandle tableHandle, SplitSchedulingStrategy splitSchedulingStrategy)
     {
         CassandraTableHandle cassandraTableHandle = (CassandraTableHandle) tableHandle;
-        List<CassandraPartition> partitions = cassandraTableHandle.getPartitions();
+
+        List<CassandraPartition> partitions;
+        String clusteringKeyPredicates;
+        if (cassandraTableHandle.getPartitions().isPresent()) {
+            partitions = cassandraTableHandle.getPartitions().get();
+            clusteringKeyPredicates = cassandraTableHandle.getClusteringKeyPredicates();
```
Comment (findepi) [inline presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraSplitManager.java]: Looks like `clusteringKeyPredicates` should be Optional too.

File: presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraSplitManager.java [hunk 5]
```diff
@@ -77,7 +92,20 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, Co
             }
         }
 
-        return new FixedSplitSource(getSplitsForPartitions(cassandraTableHandle, partitions, cassandraTableHandle.getClusteringKeyPredicates()));
+        return new FixedSplitSource(getSplitsForPartitions(cassandraTableHandle, partitions, clusteringKeyPredicates));
+    }
+
+    private static String extractClusteringKeyPredicates(CassandraPartitionResult partitionResult, CassandraTableHandle tableHandle, CassandraSession session)
```
Comment (findepi) [inline presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraSplitManager.java]: Looks similar to the logic in `io.prestosql.plugin.cassandra.CassandraMetadata#applyFilter`. Can we have a single method that does this?

File: presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java [hunk 6]
```diff
@@ -219,17 +219,20 @@ public boolean usesLegacyTableLayouts()
             unenforcedConstraint = clusteringPredicatesExtractor.getUnenforcedConstraints();
         }
 
-        if (handle.getPartitions().containsAll(partitionResult.getPartitions()) && handle.getClusteringKeyPredicates().equals(clusteringKeyPredicates)) {
+        Optional<List<CassandraPartition>> currentPartitions = handle.getPartitions();
+        if (currentPartitions.isPresent() &&
+                currentPartitions.get().containsAll(partitionResult.getPartitions()) &&
```
Comment (findepi) [inline presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java]: We used to have `containsAll`, because we misleadingly initialized `handle.getPartitions()` to empty list.
Now, this should be `! equals`.

File: presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java [hunk 7]
```diff
@@ -219,17 +219,20 @@ public boolean usesLegacyTableLayouts()
...
+                handle.getClusteringKeyPredicates().equals(clusteringKeyPredicates)) {
             return Optional.empty();
         }
 
         return Optional.of(
                 new ConstraintApplicationResult<>(new CassandraTableHandle(
                         handle.getSchemaName(),
                         handle.getTableName(),
-                        partitionResult.getPartitions(),
+                        Optional.of(partitionResult.getPartitions()),
```
Comment (findepi) [inline presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java]: Should this be an intersection of `partitionResult.getPartitions()` and `currentPartitions` (if present)?
In `io.prestosql.plugin.cassandra.CassandraPartitionManager#getPartitions` i don't see how yet how subsequent call to applyFilter should not be able to _widen_ (instead of _narrowing_) the list of partitions we're querying.

@ebyhr can you double check me on Cassandra changes

File: presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java [hunk 8]
```diff
@@ -219,17 +219,20 @@ public boolean usesLegacyTableLayouts()
...
                         clusteringKeyPredicates),
-                unenforcedConstraint));
+                        unenforcedConstraint));
```
Comment (findepi) [inline presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java]: `Load list of partitions when no constraint was applied` commit

- it appears second on Github, but should be before "Skip" commit, as this is a prerequisite change.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java [hunk 9]
```diff
@@ -177,6 +178,10 @@ private boolean arePlansSame(FilterNode filter, TableScanNode tableScan, PlanNod
                             domainTranslator.toPredicate(newDomain.simplify().transform(assignments::get))));
             constraint = new Constraint(newDomain, evaluator::isCandidate);
         }
+        else if (newDomain.isAll() && !usesLegacyTableLayouts) {
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java]: > By adding the check here, the flow is more confusing and harder to reason about.

True. I too independently added it here, because here we know whether we have evaluator....
but we can later check `contraint.predicate().isPresent()` too.

File: presto-tests/src/test/java/io/prestosql/tests/TestTpchDistributedQueries.java [hunk 10]
```diff
@@ -48,22 +46,7 @@ public void testIOExplain()
         MaterializedResult result = computeActual("EXPLAIN (TYPE IO, FORMAT JSON) " + query);
         IoPlanPrinter.IoPlan.TableColumnInfo input = new IoPlanPrinter.IoPlan.TableColumnInfo(
                 new CatalogSchemaTableName("tpch", "sf0.01", "orders"),
-                ImmutableSet.of(
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-tests/src/test/java/io/prestosql/tests/TestTpchDistributedQueries.java]: Good catch. As i understand, `applyFilter` was called only when there was some Filter. So this lack of "pull-up" is pre-existing. 
But the change here also means we could have just made it worse and it may bite in practice.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java]: Moved to `if (!metadata.usesLegacyTableLayouts(session, node.getTable())) {` branch

File: presto-tests/src/test/java/io/prestosql/tests/TestTpchDistributedQueries.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-tests/src/test/java/io/prestosql/tests/TestTpchDistributedQueries.java]: > Hmm... this is potentially a problem. Technically, the connector should still be able to describe the constraints of a given table, regardless of whether a filter was "pushed in" or not. 

@martint this is already an issue. Please see: https://github.com/prestosql/presto/issues/1399. Currently detailed predicate (e.g. involving partitions) is not returned back to planner as `TableScanNode#enforcedConstraint` is derived from pushed down predicate.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java (hunk 2 above)
Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java]: discussed offline

File: presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java (hunk 6 above)
Comment (sopel39) [reply presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java]: I didn't want to change existing behavior. Added todo

File: presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraSplitManager.java (hunk 5 above)
Comment (sopel39) [reply presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraSplitManager.java]: Tried that, but `CassandraClusteringPredicatesExtractor` is used to get also constraint in  `CassandraMetadata#applyFilter`, so use case is different

File: presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraSplitManager.java (hunk 4 above)
Comment (sopel39) [reply presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraSplitManager.java]: Let's keep it minimal

File: presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java (hunk 6 above)
Comment (findepi) [reply presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java]: > I didn't want to change existing behavior. Added todo

where's the TODO?

File: presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java [hunk 11]
```diff
@@ -221,6 +221,7 @@ public boolean usesLegacyTableLayouts()
 
         Optional<List<CassandraPartition>> currentPartitions = handle.getPartitions();
         if (currentPartitions.isPresent() &&
+                // TODO: we should skip only when new table handle does not narrow down enforced predicate
```
Comment (findepi) [inline presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java]: Squash into `Load list of partitions when no constraint was applied` commit

File: presto-tpch/src/main/java/io/prestosql/plugin/tpch/TpchMetadata.java [hunk 12]
```diff
@@ -197,17 +195,6 @@ public ConnectorTableHandle getTableHandleForStatisticsCollection(ConnectorSessi
                 .collect(toSet());
     }
 
-    @Override
-    public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTableLayoutHandle handle)
```
Comment (findepi) [inline presto-tpch/src/main/java/io/prestosql/plugin/tpch/TpchMetadata.java]: Should this be part of the `Return TPCH table predicate when no applyFilter was called` commit?

Looks like a removal of obsolete method after `Remove table layouts from TPCH connector` commit.
If correct, extract to obsolete commit
```
Remove obsolete override

TPCH doesn't use legacy TableLayouts since c1bdbd7149f35e11abec7c48b964d1e765e3c205.
```

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java [hunk 13]
```diff
@@ -186,6 +186,11 @@ private boolean arePlansSame(FilterNode filter, TableScanNode tableScan, PlanNod
         TableHandle newTable;
         TupleDomain<ColumnHandle> remainingFilter;
         if (!metadata.usesLegacyTableLayouts(session, node.getTable())) {
+            if (!constraint.predicate().isPresent() && newDomain.contains(node.getEnforcedConstraint())) {
+                // new domain is wider than domain already provided by table scan
+                return Optional.empty();
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java]: Shall we eliminate Filter in such a case?

(Consider adding a TODO for now, unless the answer is obvious.)
